### PR TITLE
feat: add `StarkScan` client to fetch transaction history

### PR DESCRIPTION
**Pull request by stanleyyconsensys**
This PR is adding a client sdk to fetch the transaction history with StarkScan

This is the first journey to enable fetching transaction by StarkScan

This PR includes:
- a starkscan data client to enable:
   - getTransactions (by address , by till to timestamp)
   - getDeployTransaction (by address)
- a dataClient interface `IDataClient` for generic interface
- a factory to create the `StarkScanClient`
- a config mapping for `StarkScanClient`

### CODE REVIEW

File: packages/starknet-snap/src/types/snapState.ts [hunk 1]
```diff
@@ -134,4 +143,26 @@ export type Transaction = {
...
+  contractAddress: string; // in hex
+  executionStatus?: TransactionExecutionStatus | string;
+  finalityStatus?: TransactionFinalityStatus | string;
+  failureReason: string;
+  timestamp: number;
+  maxFee?: string | null;
+  actualFee?: string | null;
+  // using Record<string, TranscationAccountCall[]> to support O(1) searching
+  accountCalls?: Record<string, TranscationAccountCall[]> | null;
+  version: 'V2';
```
Comment (khanti42) [inline packages/starknet-snap/src/types/snapState.ts:160]: What about V3 transactions ? 

https://starkscan.readme.io/reference/transaction-object

File: packages/starknet-snap/snap.config.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ const config: SnapConfig = {
     SNAP_ENV: process.env.SNAP_ENV ?? 'prod',
     VOYAGER_API_KEY: process.env.VOYAGER_API_KEY ?? '',
```
Comment (khanti42) [inline packages/starknet-snap/snap.config.ts:16]: When will we discontinue VOYAGER ?

File: packages/starknet-snap/src/chain/data-client/starkscan.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+      senderAddress: this.getSenderAddress(tx),
+      timestamp,
+      finalityStatus,
+      executionStatus,
+      maxFee,
+      actualFee,
+      contractAddress: this.getContractAddress(tx),
+      accountCalls,
+      failureReason: failureReason ?? '',
+      version: 'V2',
```
Comment (khanti42) [inline packages/starknet-snap/src/chain/data-client/starkscan.ts:234]: Why don't we include version from the response here ?

**Review by khanti42 — CHANGES_REQUESTED** [review]
Some clarifying comments

File: packages/starknet-snap/src/chain/data-client/starkscan.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+  /**
+   * Fetches the deploy transaction for a given contract address.
+   *
+   * @param address - The address of the contract to fetch the deploy transaction for.
+   * @returns A Promise that resolve the Transaction object or null if the transaction can not be found.
+   */
+  async getDeployTransaction(address: string): Promise<Transaction | null> {
+    // Fetch the first 5 transactions to find the deploy transaction
+    // The deploy transaction usually is the first transaction from the list
```
[the remark this answers is not in the document]
Comment (stanleyyconsensys) [reply packages/starknet-snap/src/chain/data-client/starkscan.ts:158]: it is for buffer, some very very legacy account, may use declare before deploy

File: packages/starknet-snap/src/types/snapState.ts (hunk 1 above)
Comment (stanleyyconsensys) [reply packages/starknet-snap/src/types/snapState.ts:160]: it is not for TX v3, it is for separate the type of the state data

File: packages/starknet-snap/snap.config.ts (hunk 2 above)
Comment (stanleyyconsensys) [reply packages/starknet-snap/snap.config.ts:16]: until the get-stransactions fully released , then we are safe to remove

File: packages/starknet-snap/src/chain/data-client/starkscan.ts (hunk 3 above)
Comment (stanleyyconsensys) [reply packages/starknet-snap/src/chain/data-client/starkscan.ts:234]: yea you are right, may be it is better to include a version of the transaction

**Review by khanti42 — APPROVED** [review]
LGTM

**Pull request merged**